ckan/views/dataset.py

```diff
@@ -893,60 +896,58 @@ def get(self, package_type: str, id: str) -> Union[Response, str]:
         )
 
 
-def follow(package_type: str, id: str) -> Response:
-    """Start following this dataset.
-    """
-    context: Context = {
-        u'user': current_user.name,
-        u'auth_user_obj': current_user
-    }
-    data_dict = {u'id': id}
+def follow(package_type: str, id: str) -> str:
```

Comment:
Perhaps it would be nice to clearly mark the "htmx view functions" so it's easy to see that these functions don't return a response but a string with the rendered html. 
For instance we could create an `@htmx` or `@for_htmx` decorator. That could do nothing for now, but in the future maybe it could check if the request has an htmx header (assuming the view only works with htmx). Or if the view can also render a full page we can define the template to be used if it's an htmx request (like this [pattern](https://github.com/spookylukey/django-htmx-patterns/blob/master/separate_partials_single_view.rst))

Or we could just add it to the docstrings, but a decorator feels like more flexible

Reply:
Speaking of this,  do you think we will have single views handling htmx and regular requests?

Reply:
We are using the following helper in our projects for new views that allow partial rendering with htmx:

```python
 def render(template: str, data: dict[str, Any] | None = None) -> str:
     """HTMX-aware version of tk.render.

     If a current request is made by HTMX, render a template with the same path from
     `htmx/` folder.
     When `HX-CKAN-Role` is present, apply it as a secondary extension. Use this 
     header if the single view can produce multiple partials.

     """
     if is_htmx_request():
         path, ext = os.path.splitext(template)

         if role := tk.request.headers.get("hx-ckan-role"):
             role = f".{role}"

         template = os.path.join("htmx", f"{path}{role}{ext}")

     return tk.render(template, data)
```

And, if you are using it in view instead of `tk.render`, the following happens when you are calling `render("group/read.html")`, for example:

* it it's a non-htmx request, `tk.render` is called as is 
* if it's a htmx request with no extra headers, `htmx/` added before the template name and we are rendering `tk.render("htmx/group/read.html")`. In this way, you can support rendering pages in the normal way and via HTMX. Basically, that's what @pdelboca is doing for the dataset search page.
* if there is an `hx-ckan-role: <ROLE>` header, we are using `<ROLE>` as a secondary extension, between the template name and template extension: `name.<ROLE>.html`. If HTMX request contains `hx-ckan-role: followers` header when requesting the group details page, the following template is rendered: `tk.render("htmx/group/read.followers.html")`. In this way, the view can update the main part of the page like in the previous point + update a number of separate segments(sidebar, header, footer)  separately.

Reply:
I would go for a more explicit @for_htmx('template.html'') since it is more readable. It's true that the if pattern will appear a lot and the solution @amercader proposes seems logic.

Making `render` aware of `HTMX` and internally handling the logic makes it less transparent to the user. Is one of those double behavior function that are not explicit in the name and can be hard to follow up. Is a function in which the behavior does not change due to parameters but rather to a global object it does not have any control over. (Event when Flask uses Global as a pattern, I'm not always comfortable with it.)

@amercader regarding your second question. Using HTMX assumes that we are Okay with the principles of returning pieces of HTML to replace elements. Once you are okay with that principle, the library recommends to be [as specific as you want with your endpoints](https://hypermedia.systems/building-a-contacts-app-with-hyperview/#_hypermedia_notes_api_endpoints) since an endpoint will exist only to solve a particular UX/UI need on a given time. (Instead of a generic `package_show` that everyone can reuse). 

I will expect with time to have both a mix of views with @for_htmx decorator (that are useful in less dynamic views where we just need to refresh content) and several specific views with HTML snippets that we will use to make the UX more component oriented.

Reply:
I like the decorator idea.

For many views it will make sense to have the same code run for htmx and non-htmx version, and this can be explicit in the view checking just before rendering.

For the htmx-only views an `@htmx_only` decorator would be nice for automatically 404ing non-htmx requests.

Reply:
What should we do with this suggestions? I like the ideas but on the other hand I'm reluctant to start implementing tools for it before we actually start using it.

Would it be a better fit for another PR?

Reply:
Let's port some more functionality to htmx and see what patterns emerge, and then it will be easier to decide for which option to settle